### PR TITLE
[DS-3532] Breadcrumbs in header

### PR DESCRIPTION
**Pull request by benbosman**
This PR provides support for breadcrumbs in header based on https://github.com/DSpace/Rest7Contract/pull/100

It exposes links to the parent community

### CODE REVIEW

**Review by tdonohue — COMMENTED** [review]
@benbosman : Looks good!  I've tested this out and it works as expected.  The responses also look to match the REST contract. ITs also look good to me.  All that said, I have a few requests for JavaDocs before we would merge this.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: I don't see any value in adding a more test for AB as this is the same test data category than AA

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: we can avoid to retrieve this value from the REST, the findOne endpoint has already test in the specific IT class. Doing it again here will only slow our test

**Review by abollini — CHANGES_REQUESTED** [review]
I recomend to simplify and revisit a bit the test reusing the existing Matchers and avoiding extra REST call that are already tested in the CommunityRestRepositoryIT or CollectionRestRepositoryIT. If some test for the basic collection / community get method are missing they should be added in the above classes

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: Good point, yes there is a `org.dspace.app.rest.matcher.CollectionMatcher` which can be used to compare JSON results against a known set of values.  There's some good examples in the `CollectionRestRepositoryIT`.

In fact, all the tests in this IT should likely be updated to use either `CollectionMatcher` or `CommunityMatcher`.  (and those two matchers could be updated with new comparison methods as needed).  I had completely overlooked that in this PR.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: we should find a better name for this IT class as it also include test related to items. Should be `TraverseCommunityCollectionItemParent` ? mabye a sentence as javadoc explaining that could be also useful

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: can we use the same name convention used for the community to make easier follow the hierarchy? colAA1, colAA2, colAB1

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: as above, maybe itemAA, itemAAmappedInAB, itemAB

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: my suggestion is to remove this test as it is functionally equivalent to col1ParentCommunityTest

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: my suggestion is to remove this test as it is functionally equivalent to col1ParentCommunityTest

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: we need all the test related to the security and wrong scenarios. 
The specified item, collection, community 
- don't exist, don't make sense (i.e. it is not an uuid) -> (404)
- are not visibile (401, 403)
- the starting object is visible but the parent is not visible (401, 403)

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CollectionMatcher.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CollectionMatcher.java]: please don't reduce the scope of the matcher as this would have the effect to reduce the usefulness of the existing tests. Instead, if a new match method that exclude embedded is required overload the method

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java]: again better overload the method than reduce the checks

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java]: again better overload the method than reduce the checks

**Review by abollini — CHANGES_REQUESTED** [review]
thanks for the update. I noted that some test scenario are missing, I'm fine to postpone security to a separate issue if needed (please open a ticket if so) but I like to have the missing test for the edge scenario in place and the feedback on the Matchers addressed before the merge

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: @abollini : I'm not sure the 404 test is needed, as we already have corresponding 404 tests in ITs for Communities, Collections & Items. How can a 404 test here add anything different?

I do agree that the other two types of tests would be worth adding, though we could consider doing so as a separate security ticket (as you noted), provided that we create a ticket in JIRA.

Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: obvious test are useful for the future because they look obvious now that we know the implementation but if we go trough a refactoring the implementation could be completely different and break something that we expect to just work. So I always recommend to make black box test, indeed we should follow a TDD approach that would imply that.

Anyway, I'm not completely sure that it will return 404 right now, maybe we can get a NPE o ClassCastException so a 500 see https://github.com/DSpace/DSpace/pull/2673/files#diff-59c10eaa99f8fbc2322484a9bacff5afR51 (again I'm just guessing but it is wrong to look to the communityservice to get the answer, we need to have test to proof it)

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: this has been addressed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionItemParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionItemParentIT.java]: we need a bug issue for that

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionParentIT.java]: it has been addressed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionItemParentIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionItemParentIT.java]: can you rename the test methods according to the refactored name of the items variable?

**Review by abollini — CHANGES_REQUESTED** [review]
other than the two small inline comments I want to highlight that the refactoring of the Matchers introducing an "overload" method named with a postfix FullProjection will reduce the effect of test in other classes. For instance I see use of the `CollectionMatcher` in the `WorkflowItemRestRepositoryIT` that is not touched by this PR (but other are also involved and we should check also for the CommunityMatcher) this mean that the existent test with this PR will check less detail about the Collection also when the fullprojection is expected because they still use the old method name that now has a different meaning. Instead than rename the existent method in the "FullProjection" version that will imply a lot of changes in other IT (with the consequence of lot of conflict in other PRs) I suggest to name the reduced matcher method in `matchCollectionEntryNoProjection` and leave untouched the existing methods

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CollectionMatcher.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CollectionMatcher.java]: I agree with @abollini's feedback here. If we want a match method which doesn't check for a logo, let's create a separate `matchCollectionEntryNoLogo()` method (or similar).

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java]: I agree here as well.  Instead of removing the checks for embedded content from `matchCommunityEntry()`, we should create a separate method named `matchCommunityEntryNoEmbeds()` (or similar) which does this.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java]: I agree here as well. I think we should *not* change the current behavior of `matchCommunityEntry()`.  Instead, we should create a new method called `matchCommunityEntryNoEmbeds()` which performs the match without embedded content.

**Review by tdonohue — COMMENTED** [review]
UPDATE:  @abollini on second review I see that the Matcher behavior changes have been updated in all other ITs (see the new changes to `CommunityRestRepositoryIT` and `CollectionRestRepositoryIT` which were not previously a part of this PR).

So, Raf chose to fix the feedback around Matchers in a different manner.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CollectionMatcher.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CollectionMatcher.java]: @abollini : This was resolved in a different manner.  A new `matchCollectionEntryFullProjection()` method was added, and all existing tests that depend on embeds (full projection) were updated to use that new method.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java]: @abollini : This was resolved in a different manner.  A new `matchCommunityEntryFullProjection()` method was added, and all existing tests that depend on embeds (full projection) were updated to use that new method.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/CommunityMatcher.java]: @abollini : This was resolved in a different manner. A new matchCommunityEntryFullProjection() method was added, and all existing tests that depend on embeds (full projection) were updated to use that new method.

**Comment by abollini** [comment]
@tdonohue not all the ITs have been refactored for instance not the `WorkflowItemRestRepositoryIT` that I have point out in the previous comment https://github.com/DSpace/DSpace/blob/9de65efa8c26d106ff97d8437872e4a016462f61/dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkflowItemRestRepositoryIT.java#L461

I would prefer to avoid a manual check, if we do this kind a refactor it should be done with the IDE... but maybe it has been used and the test has been introduced in a more recent PR, anyway this show how risky is this approach.
All the existing PR will get a conflict or, if no conflict is raised will change their expected behavior after that this PR will be merged.
We should not reuse an existing method to do something different

**Comment by tdonohue** [comment]
@abollini : yes, thanks for highlighting that. I just noticed that myself after looking at code in my IDE.

It looks like (to me) that all `CommunityMatcher` tests have been properly updated (they all seem to be in `CommunityRestRepositoryIT`. 

However, there's a number of `CollectionMatcher` tests (using `projection=full`) that were not updated in this PR.  Here's the ITs where I see them:
* ItemOwningCollectionUpdateRestControllerIT
* MappedCollectionRestRepositoryIT
* WorkflowItemRestRepositoryIT 
* WorkspaceItemRestRepositoryIT

So, this PR should either revert the Matcher changes (as @abollini is suggesting) or update *all the ITs* to correct them to use the new `FullProjection` methods whenever `projection=full` is passed as a param.

**Comment by abollini** [comment]
I'm fine with both solution for **this specific** PR, keep in mind that update all the ITs  will add more work on coming PRs as they will have conflict to solve and we should recheck if new test should use the new FullProjection method instead than the default one.
Most important, I really like to make the point to pay attention to semantic change in the meaning of methods / services because this have wide impact so they should be avoid

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionItemParentIT.java
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityCollectionItemParentIT.java]: https://jira.lyrasis.org/browse/DS-4444

**Review by tdonohue — APPROVED** [review]
👍 Re-reviewed and re-tested today. This looks good to me, and it resolves (refactors) all existing ITs based on the `CollectionMatcher` and `CommunityMatcher` changes (as requested by @abollini).

**Comment by abollini** [comment]
well, it is not what we have agreed as far as I understand but if other are fine with that I turn my -1 in a 0.
The test have been not refactored to use the new name, instead they have been changed to don't check the projection anymore, see for instance
https://github.com/DSpace/DSpace/pull/2673/commits/2a54bfab4015afe1b3075eaff692dbb0c4facc3b#diff-3e35159035303e83774c35edbff256d9R101
so this PR is changing the existing test behavior. Again, assuming that it is only a "style" decision and there are no issue in run the original test it can go also in this way

**Comment by tdonohue** [comment]
I've double checked all the ITs in this PR that remove `projections=full`.   In all those cases, projections are not necessary.  I think the `projections=full` was just added as a temporary fix to allow these tests to work properly in the PR where projections were added.

**Pull request merged**
